File: internal/model/shop.go
package model

import (
	"time"

	"gorm.io/gorm"
)

type ShopItem struct {
	ID          string         `gorm:"primaryKey;size:36" json:"id"`
	Name        string         `gorm:"size:128;not null" json:"name"`
	Description string         `gorm:"size:512" json:"description"`
	ImageURL    string         `gorm:"size:512" json:"image_url"`
	PointsCost  int            `gorm:"not null;default:0" json:"points_cost"`
	Stock       int            `gorm:"not null;default:-1" json:"stock"`
	IsActive    bool           `gorm:"not null;default:true" json:"is_active"`
	CreatedAt   time.Time      `json:"created_at"`
	UpdatedAt   time.Time      `json:"updated_at"`
	DeletedAt   gorm.DeletedAt `gorm:"index" json:"-"`
}

func (ShopItem) TableName() string { return "shop_items" }

type Redemption struct {
	ID          string    `gorm:"primaryKey;size:36" json:"id"`
	UserID      string    `gorm:"size:36;not null;index" json:"user_id"`
	ItemID      string    `gorm:"size:36;not null" json:"item_id"`
	ItemName    string    `gorm:"size:128" json:"item_name"`
	PointsSpent int       `gorm:"not null" json:"points_spent"`
	Status      string    `gorm:"size:32;default:completed" json:"status"`
	CreatedAt   time.Time `json:"created_at"`
}

func (Redemption) TableName() string { return "redemptions" }

File: web/src/api/modules/shop.ts
import { api } from '@/api/client'
import type { ApiResponse } from '@/types'

export interface ShopItem {
  id: string
  name: string
  description: string
  image_url: string
  points_cost: number
  stock: number
  is_active: boolean
  created_at: string
  updated_at: string
}

export interface Redemption {
  id: string
  user_id: string
  item_id: string
  item_name: string
  points_spent: number
  status: string
  created_at: string
}

export interface RedeemResult {
  redemption: Redemption
  user_points: number
}

export function getShopItems() {
  return api.get<ApiResponse<ShopItem[]>>('/shop/items')
}

export function redeemItem(itemId: string) {
  return api.post<ApiResponse<RedeemResult>>('/shop/redeem', { item_id: itemId })
}

export function getHistory(page = 1, pageSize = 20) {
  return api.get<ApiResponse<{ items: Redemption[]; pagination: { page: number; total: number } }>>('/shop/history', {
    page,
    page_size: pageSize,
  })
}

export function createShopItem(data: Omit<ShopItem, 'id' | 'created_at' | 'updated_at'> & { stock?: number }) {
  return api.post<ApiResponse<ShopItem>>('/admin/shop/items', data)
}

export function updateShopItem(id: string, data: Omit<ShopItem, 'id' | 'created_at' | 'updated_at'> & { is_active: boolean; stock?: number }) {
  return api.put<ApiResponse<ShopItem>>(`/admin/shop/items/${id}`, data)
}

export function deleteShopItem(id: string) {
  return api.delete(`/admin/shop/items/${id}`)
}
